pydantic_ai_slim/pydantic_ai/ag_ui.py

```diff
@@ -0,0 +1,871 @@
+    _logger: logging.Logger = field(default=_LOGGER, repr=False, init=False)
+
+    async def run(
+        self,
+        run_input: RunAgentInput,
+        accept: str = SSE_CONTENT_TYPE,
+        *,
+        output_type: OutputSpec[RunOutputDataT] | None = None,
+        model: Model | KnownModelName | str | None = None,
+        deps: AgentDepsT = None,
+        model_settings: ModelSettings | None = None,
+        usage_limits: UsageLimits | None = None,
+        usage: Usage | None = None,
+        infer_name: bool = True,
+        toolsets: Sequence[AbstractToolset[AgentDepsT]] | None = None,
+    ) -> AsyncGenerator[str, None]:
+        """Run the agent with streaming response using AG-UI protocol events.
+
+        The first two arguments are specific to `Adapter` the rest map directly to the `Agent.iter` method.
+
+        Args:
+            run_input: The AG-UI run input containing thread_id, run_id, messages, etc.
+            accept: The accept header value for the run.
+
+            output_type: Custom output type to use for this run, `output_type` may only be used if the agent has no
+                output validators since output validators would expect an argument that matches the agent's output type.
+            model: Optional model to use for this run, required if `model` was not set when creating the agent.
+            deps: Optional dependencies to use for this run.
+            model_settings: Optional settings to use for this model's request.
+            usage_limits: Optional limits on model request count or token usage.
+            usage: Optional usage to start with, useful for resuming a conversation or agents used in tools.
+            infer_name: Whether to try to infer the agent name from the call frame if it's not set.
+            toolsets: Optional list of toolsets to use for this agent, defaults to the agent's toolset.
+
+        Yields:
+            Streaming SSE-formatted event chunks.
+        """
+        self._logger.debug('starting run: %s', json.dumps(run_input.model_dump(), indent=2))
+
+        encoder: EventEncoder = EventEncoder(accept=accept)
+        if run_input.tools:
+            # AG-UI tools can't be prefixed as that would result in a mismatch between the tool names in the
+            # PydanticAI events and actual AG-UI tool names, preventing the tool from being called. If any
+            # conflicts arise, the AG-UI tool should be renamed or a `PrefixedToolset` used for local toolsets.
+            toolset: AbstractToolset[AgentDepsT] = DeferredToolset[AgentDepsT](
+                [
+                    ToolDefinition(
+                        name=tool.name,
+                        description=tool.description,
+                        parameters_json_schema=tool.parameters,
+                    )
+                    for tool in run_input.tools
+                ]
+            )
+            toolsets = [toolset] if toolsets is None else [toolset] + list(toolsets)
```

Comment:
The `iter` `toolsets` _override_ `toolsets` passed to `Agent.__init__`, which would be unexpected in this context (as not passing any toolsets to `to_ag_ui` would result in all non-AGUI toolsets being dropped). So we should include `agent._user_toolsets` here (and I should probably make that a public method!)

Also I'm not sure this is actually better, but it does look nicer:

```suggestion
            toolsets = [toolset, *(toolsets or [])]
```